modules/core/src/main/java/org/apache/ignite/lang/IgniteLogger.java

```diff
@@ -195,7 +201,12 @@ private void logInternal(Level level, String msg, Object... params) {
         if (!log.isLoggable(level))
             return;
 
-        log.log(level, LoggerMessageHelper.arrayFormat(msg, params));
+        Throwable throwable = LoggerMessageHelper.getThrowableCandidate(params);
+
+        if (throwable != null)
+            log.log(level, LoggerMessageHelper.arrayFormat(msg, LoggerMessageHelper.trimmedCopy(params)), throwable);
+        else
+            log.log(level, LoggerMessageHelper.arrayFormat(msg, params));
```

Comment:
From my side, this part has incorrect logic. The code is not broken, but the idea is wrong.
Using vararg for things of a different kind is a bad practice.
I believe method contract MUST declate it explicitly like in log(Level, String, Throwable).

Lets me explain:
1. Usability. Here and in any other logger, "params" argument is a list of parameters for a message template.
Vararg itself is ok here while it contains only params.
Generally, vararg is already hard to use, because it is very easy to miss one or write the extra one.
Throwable as a last optional argument makes things ever worse.
What if the user passes Throwable as a param? or passes the extra argument? 

2. Compatibility. Slf4J hadn't supported this case. So, no compatibility is broken.

3. Codestyle. Having an implicit argument is a bad style.
Why you don't declare log(Object... args) and cast first arg to Level, second to String?
This top-level public class, which is widely used in the project, shouldn't push developers for writing uncertain code.

4. Performance. If you bother with extra Supplier<String> object creation, then vararg is actually array and extra object (array) will be created anyway + jvm will need to fill the array.
So, you win no nanosecond here and maybe even loose.

Reply:
Possible solutions are
- to drop and force a developer to format the message before the method call or pass Supplier<String> for message.
- rewrite with log(Level, String, Throwable, Object...)

Reply:
> Compatibility. Slf4J hadn't supported this case. So, no compatibility is broken.
I couldn't agree with you. The only reason why I decided to support such methods is that Slf4J supports throwable as the last param [1] and it is widely used in Jraft module (for example [2]). So, most logic with checking the last param was taken from Slf4J library [3]. As far as I can understand, the current API was changed by @vldpyatkov and it is a subset of Slf4J API. We use `LoggerMessageHelper` [4] and this class is very close to `MessageFormatter` from Slf4J [5]

So, I would say, that my implementation should be approved, as far as it preserves logic from Slf4J

[1] https://github.com/qos-ch/slf4j/blob/975a9a1e7e15e1e9ab8b34487aecba15048a7ded/slf4j-api/src/main/java/org/slf4j/helpers/AbstractLogger.java#L397
[2] https://github.com/apache/ignite-3/blob/96067912780385940b3db432f3af1d89531233cf/modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/impl/LocalLogStorage.java#L229
[3] https://github.com/qos-ch/slf4j/blob/master/slf4j-api/src/main/java/org/slf4j/helpers/MessageFormatter.java#L155
[4] https://github.com/apache/ignite-3/blob/96067912780385940b3db432f3af1d89531233cf/modules/core/src/main/java/org/apache/ignite/lang/LoggerMessageHelper.java#L95
[5] https://github.com/qos-ch/slf4j/blob/975a9a1e7e15e1e9ab8b34487aecba15048a7ded/slf4j-api/src/main/java/org/slf4j/helpers/MessageFormatter.java#L178

Reply:
> I couldn't agree with you. The only reason why I decided to support such methods is that Slf4J supports throwable as the last param [1]

log() methods with vararg in org.slf4j.Logger interface has no Throwable mentioned. 
It's sad that they have a bad API. Do we want to be like them?

However, they refer to a FAQ page on their website [1], where said your way is possible.
They said that _If the exception is not the last argument, it will be treated as a plain object and its stack trace will NOT be printed._ I think it is critical to have a thread dump when smth goes wrong despite developer's mistakes.
But their next _"However, such situations should not occur in practice."_, sorry, is bullshit.

>  it is widely used in Jraft module (for example [2])
We already incorporated JRaft code into Ignite module, so the correct way is to fix this.
Or at least, create a separate JRaftLogger (as a wrapper) in JRaft module
rather than go error-prone way.

This just means, SLF4J is not the ideal we have to look at when making long-term decisions.

[1] http://www.slf4j.org/faq.html#paramException